[CodeGen] Avoid sinking vector comparisons during CodeGenPrepare

david-arm opened the pull request:
[CodeGen] Avoid sinking vector comparisons during CodeGenPrepare

Whilst reviewing PR #109289 and doing some analysis with various
tests involving predicated blocks I noticed that we're making
codegen and performance worse by sinking vector comparisons
multiple times into blocks. It looks like the sinkCmpExpression
in CodeGenPrepare was written for scalar comparisons where there
is only a single condition register, whereas vector comparisons
typically produce a vector result. For some targets, such a NEON
or SVE, there are multiple allocatable vector registers that can
store the result and so we should avoid sinking in that case.

goldsteinn commented on llvm/lib/CodeGen/CodeGenPrepare.cpp:
```diff
@@ -1779,29 +1779,35 @@ static bool sinkCmpExpression(CmpInst *Cmp, const TargetLowering &TLI) {
+  SmallSet<User *, 4> Users;
+  for (auto *U : Cmp->users()) {
+    Instruction *User = cast<Instruction>(U);
+    if (isa<PHINode>(User))
```
Maybe keep some full-user iteration below and also skip pseudo instructions that will be dropped during codegen here?

goldsteinn commented:
> Alternatively, I could also introduce a TLI hook to query the
> target if this is a preferred solution?

This seems pretty similiar to existing `shouldNormalizeToSelectSequence` or `hasMultipleConditionRegisters`.
Maybe `shouldNormalizeToSelectSequence` should become `shouldSinkComparisons` more generally (whether it be into selects or BBs).

davemgreen commented:
MVE has only a single vector predicate register available, so it might be better to do the sinking. Could you add a test case to check if it looks better?

david-arm commented:
> MVE has only a single vector predicate register available, so it might be better to do the sinking. Could you add a test case to check if it looks better?

That's a fair point. I thought this was being tested by existing tests in llvm/test/CodeGen/Thumb2/LowOverheadLoops/fast-fp-loops.ll. I took a look at the results and they seemed to be around 1-2 instructions less overall, but that doesn't mean it's better of course. :) After chatting privately with @paulwalker-arm perhaps it's better to introduce a `hasSingleVectorComparisonRegister` or something like that? For NEON, SVE and presumably AVX this returns a number greater than one. For other targets where the hook returns 1 we sink the comparison. I'll play around with this and see where it leads ...

david-arm commented:
Hi @davemgreen, I will also add a proper test for MVE similar to the one I added because I realised the existing Thumb2 test changes are for tail-folded loops, which is probably a bit different.

david-arm commented:
> > Alternatively, I could also introduce a TLI hook to query the
> > target if this is a preferred solution?
> 
> This seems pretty similiar to existing `shouldNormalizeToSelectSequence` or `hasMultipleConditionRegisters`. Maybe `shouldNormalizeToSelectSequence` should become `shouldSinkComparisons` more generally (whether it be into selects or BBs).

Hmm, it looks like `shouldNormalizeToSelectSequence` is a virtual function overridden by AArch64 and is used by the DAGCombiner for a different problem. It looks like changing this function would be beyond the scope of this patch, however I am happy to investigate this as a follow-on if that's ok? Is the idea to avoid performing the DAG combine too if for vector comparisons the target has multiple vector predicate registers?

goldsteinn commented:
> > > Alternatively, I could also introduce a TLI hook to query the
> > > target if this is a preferred solution?
> > 
> > 
> > This seems pretty similiar to existing `shouldNormalizeToSelectSequence` or `hasMultipleConditionRegisters`. Maybe `shouldNormalizeToSelectSequence` should become `shouldSinkComparisons` more generally (whether it be into selects or BBs).
> 
> Hmm, it looks like `shouldNormalizeToSelectSequence` is a virtual function overridden by AArch64 and is used by the DAGCombiner for a different problem. It looks like changing this function would be beyond the scope of this patch, however I am happy to investigate this as a follow-on if that's ok? Is the idea to avoid performing the DAG combine too if for vector comparisons the target has multiple vector predicate registers?

Sure, moreso what I meant was it seems we essentially have 2 existing flags that are related to avoiding transforms targeting single-predicate cmp ISAs. If you do choose to add a flags, it probably should co-opt or replace one of those.

goldsteinn commented on llvm/include/llvm/CodeGen/TargetLowering.h:
```diff
@@ -502,6 +502,11 @@ class TargetLoweringBase {
+  /// Return true if multiple vector predicate registers are available.
+  bool hasMultipleVectorPredicateRegisters() const {
+    return HasMultipleVectorPredicateRegisters;
+  }
```
Imo this should just take a type and be `hasMultiplePredicateRegister(Type * Ty)`

jayfoad commented on llvm/include/llvm/CodeGen/TargetLowering.h:
```diff
@@ -2389,7 +2389,7 @@ class TargetLoweringBase {
     // If a target has multiple condition registers, then it likely has logical
     // operations on those registers.
-    if (hasMultipleConditionRegisters())
+    if (hasMultiplePredicateRegisters(VT))
```
From looking at the use of this hook in `DAGCombiner::visitSELECT`, I think VT here is the type of X and Y. But `hasMultiplePredicateRegisters` should be passed the type of the predicate, which is hard coded to `MVT::i1`. (Please double check this since I find it quite confusing!)

jayfoad replied:
Of course it would help if the comment on this function said what `VT` is supposed to be.

jayfoad commented on llvm/lib/Target/AMDGPU/AMDGPUISelLowering.h:
```diff
@@ -387,6 +387,16 @@ class AMDGPUTargetLowering : public TargetLowering {
   }
+
+  virtual bool hasMultiplePredicateRegisters(EVT VT) const override {
+    // FIXME: This is only partially true. If we have to do vector compares,
```
AMDGPU part looks fine. In future we will probably want to address this FIXME by passing in a node, not just a type.

davemgreen commented on llvm/lib/Target/AArch64/AArch64ISelLowering.h:
```diff
@@ -1358,6 +1358,10 @@ class AArch64TargetLowering : public TargetLowering {
 
   bool softPromoteHalfType() const override { return true; }
+
+  virtual bool hasMultiplePredicateRegisters(EVT VT) const {
```
+override

davemgreen commented on llvm/test/CodeGen/Thumb2/mve-sink-vector-cmp.ll:
```diff
@@ -0,0 +1,136 @@
+vector.body:
+  %index = phi i64 [ 0, %entry ], [ %index.next, %pred.store.continue18 ]
+  %vec.ind = phi <4 x i64> [ <i64 0, i64 1, i64 2, i64 3>, %entry ], [ %vec.ind.next, %pred.store.continue18 ]
+  %0 = icmp ult <4 x i64> %vec.ind, <i64 15, i64 15, i64 15, i64 15>
```
Can you make this 4 x i32, so that it is a legal type under MVE.

david-arm replied:
Sure, although sadly this is genuinely what the loop vectoriser kicks out, which is why I wrote the test this way. It would be good to fix the vectoriser to choose a more sensible vector induction variable when you know the max range of possible element values.

david-arm replied:
Oh of course you're right. I think I took the output from the vectoriser for AArch64 and then tried to recreate a MVE test from it. If I compile

```
void foo(int *dest) {
  for (int i = 0; i < 14; i++) {
    dest[i] = 3;
  }
}
```

with

```
clang -O3 -fno-unroll-loops --target=thumb-linux-gnueabi -mcpu=cortex-m55 -mllvm -tail-predication=disabled -S -emit-llvm ./lowtrip.c
```

then I see

```
vector.body:                                      ; preds = %vector.body, %entry
  %index = phi i32 [ 0, %entry ], [ %index.next, %vector.body ]
  %broadcast.splatinsert = insertelement <4 x i32> poison, i32 %index, i64 0
  %broadcast.splat = shufflevector <4 x i32> %broadcast.splatinsert, <4 x i32> poison, <4 x i32> zeroinitializer
  %vec.iv = or disjoint <4 x i32> %broadcast.splat, <i32 0, i32 1, i32 2, i32 3>
  %0 = icmp ult <4 x i32> %vec.iv, <i32 14, i32 14, i32 14, i32 14>
  %1 = getelementptr inbounds i32, ptr %dest, i32 %index
  tail call void @llvm.masked.store.v4i32.p0(<4 x i32> <i32 3, i32 3, i32 3, i32 3>, ptr %1, i32 4, <4 x i1> %0), !tbaa !5
  %index.next = add i32 %index, 4
  %2 = icmp eq i32 %index.next, 16
  br i1 %2, label %for.cond.cleanup, label %vector.body, !llvm.loop !9
```

I think a better example where it goes wrong is:

```
void foo(short *dest) {
  for (int i = 0; i < 14; i++) {
    dest[i] = 3;
  }
}
```

where the vectoriser output is:

```
vector.body:                                      ; preds = %vector.body, %entry
  %index = phi i32 [ 0, %entry ], [ %index.next, %vector.body ]
  %broadcast.splatinsert = insertelement <8 x i32> poison, i32 %index, i64 0
  %broadcast.splat = shufflevector <8 x i32> %broadcast.splatinsert, <8 x i32> poison, <8 x i32> zeroinitializer
  %vec.iv = or disjoint <8 x i32> %broadcast.splat, <i32 0, i32 1, i32 2, i32 3, i32 4, i32 5, i32 6, i32 7>
  %0 = icmp ult <8 x i32> %vec.iv, <i32 14, i32 14, i32 14, i32 14, i32 14, i32 14, i32 14, i32 14>
  %1 = getelementptr inbounds i16, ptr %dest, i32 %index
  tail call void @llvm.masked.store.v8i16.p0(<8 x i16> <i16 3, i16 3, i16 3, i16 3, i16 3, i16 3, i16 3, i16 3>, ptr %1, i32 2, <8 x i1> %0), !tbaa !5
  %index.next = add i32 %index, 8
  %2 = icmp eq i32 %index.next, 16
  br i1 %2, label %for.cond.cleanup, label %vector.body, !llvm.loop !9
```

and this time the <8 x i32> isn't ideal - we could in fact have chosen <8 x i16>. Anyway, for the purpose of this particular PR it doesn't matter I suppose. I'll amend the test to use i32!

davemgreen commented on llvm/test/CodeGen/Thumb2/mve-sink-vector-cmp.ll:
```diff
@@ -0,0 +1,136 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple=thumbv8.1m.main -mattr=+mve.fp,+fp-armv8d16sp,+fp16,+fullfp16 < %s | FileCheck %s
```
-mattr=+mve I think should be enough

davemgreen reviewed (commented):
Can you add these two tests too: https://godbolt.org/z/G1Ma7Gd9v. They are simpler and don't involve scalarization of the predicate (which will hopefully be rare).

david-arm commented:
> Can you add these two tests too: https://godbolt.org/z/G1Ma7Gd9v. They are simpler and don't involve scalarization of the predicate (which will hopefully be rare).

Damn, just realised I missed this. I'll take a look now.

david-arm committed:
Add tests

david-arm committed:
[CodeGen] Avoid sinking vector comparisons during CodeGenPrepare

Whilst reviewing PR #109289 and doing some analysis with various
tests involving predicated blocks I noticed that we're making
codegen and performance worse by sinking vector comparisons
multiple times into blocks. It looks like the sinkCmpExpression
in CodeGenPrepare was written for scalar comparisons where there
is only a single condition register, whereas vector comparisons
typically produce a vector result. For some targets, such a NEON
or SVE, there are multiple allocatable vector registers that can
store the result and so we should avoid sinking in that case.

david-arm commented:
Added new tests

davemgreen reviewed (commented):
Sorry for the delay I was going through things in a strange order. I had previously only checked the test on AArch64. Do we need to worry about the extra xtn/sshll this introduces? I was going to suggest we just remove them post-isel, but that might require getting the top bits correct which requires a lot of infrastructure that currently doesn't exist.

david-arm commented:
> Sorry for the delay I was going through things in a strange order. I had previously only checked the test on AArch64. Do we need to worry about the extra xtn/sshll this introduces? I was going to suggest we just remove them post-isel, but that might require getting the top bits correct which requires a lot of infrastructure that currently doesn't exist.

Do you mean for the `no_sink_simple` case? So the typical use case that we see here is when vectorising a loop with a condition and having predicated blocks that depend upon each element of the result. So when there are 4 or more uses to sink (e.g. `vector_loop_with_icmp`) this definitely gives some significant performance gains, which was the motivation for this patch. You could be right that this isn't worth it when there are only two uses, but I'm not sure how to account for that. If possible I'd prefer to keep the `hasMultiplePredicateRegisters` function as it is because I think it's more useful as a standalone hook rather than being tied to the specifics of sinking a comparison instruction. It looks like `sinkCmpExpression` was never originally intended for sinking vector instructions, and presumably this just started happening by accident for vectors without necessarily doing a cost-benefit analysis. The only clean way I can think of to work around this is by introducing a brand new TLI hook also called `shouldConsiderSinkingCmp` that checks if the number of uses of a fixed-width vector are > 2, e.g.

```
static bool sinkCmpExpression(const DataLayout &DL, CmpInst *Cmp,
                              const TargetLowering &TLI) {
  EVT ResVT = TLI.getValueType(DL, Cmp->getType());
  if (!TLI.shouldConsiderSinkingCmp(Cmp) || TLI.hasMultiplePredicateRegisters(ResVT))
    return false;

  // Avoid sinking soft-FP comparisons, since this can move them into a loop.
  if (TLI.useSoftFloat() && isa<FCmpInst>(Cmp))
    return false;
```

where shouldConsiderSinkingCmp returns true in the default case and for AArch64 does something like this:

```
bool shouldConsiderSinkingCmp(CmpInst *Cmp) {
  auto VecTy = dyn_cast<FixedVectorType>(Cmp->getType());
  return !VecTy || Cmp->getNumUsers() > 2;
}
```

What do you think? Or it could go one step further and look at the users to see if they are extractelement or not, i.e. return false if and only if it's a 2-element vector where each user is an extract?

davemgreen commented:
I believe it is less about the number of uses and more about the type legalization of the uses. v4i1 gets legalized to v4i16, which will insert xtn and sshll if the original instructions were v4i32. The cross-basic block nature makes it difficult for DAG to deal with. If the uses are extracts then the problem is lessened as extracts can always use the type it sees.

I gave it a test though and it seemed OK, so LGTM. We might want to look at removing the legalization artefacts too.

paulwalker-arm reviewed (commented):
Hi @david-arm, sorry for the late review but when looking at the tests I think I've misunderstood our original conversation.  While `hasMultipleConditionRegisters` seems like an overly specific question for the affected code, on reflection I'm not sure that matters.

Looking at `vector_loop_with_icmp` I can see that not sinking the compare improves the output, but isn't the original output only bad because the code generator is missing an obvious `extractelt (setcc X, C), IndexC => setcc (extractelt X, IndexC), C` combine? which would be better all-round.

david-arm commented:
> Hi @david-arm, sorry for the late review but when looking at the tests I think I've misunderstood our original conversation. While `hasMultipleConditionRegisters` seems like an overly specific question for the affected code, on reflection I'm not sure that matters.
> 
> Looking at `vector_loop_with_icmp` I can see that not sinking the compare improves the output, but isn't the original output only bad because the code generator is missing an obvious `extractelt (setcc X, C), IndexC => setcc (extractelt X, IndexC), C` combine? which would be better all-round.

I'm happy to have a look at your suggestion and I can believe it helps, but not sure that in general it's a good idea to be sinking unless you can prove the vector comparison will disappear? Anyway, I can try the codegen first and see if this patch still helps on top of it.

david-arm commented:
I've followed the suggestion by @paulwalker-arm and I will abandon this patch in favour of https://github.com/llvm/llvm-project/pull/116031
